compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEAssignNpuDmaBdIds.cpp

```diff
@@ -75,186 +75,507 @@ std::optional<uint32_t> getNumberIterations(scf::ForOp loop) {
-    // If the BD ID is a semi-affine expression.
-    if (bdIdOpToBdIdsMap.contains(bdIdOp)) {
-      for (uint32_t bdId : bdIdOpToBdIdsMap[bdIdOp]) {
-        generator.releaseBdId(bdId);
+    if (!shimTileToGeneratorMap.contains(tileOp.getResult())) {
+      return bdIdOp.emitOpError()
+             << "no BD ID generator found for this BD ID op's tile";
+    }
+
+    ChannelBdIdGenerator &generator =
+        shimTileToGeneratorMap[tileOp.getResult()];
+    Value value = bdIdOp.getValue();
+    if (auto op = value.getDefiningOp<affine::AffineApplyOp>()) {
+      // If the BD ID is a semi-affine expression.
+      if (bdIdOpToBdIdsMap.contains(bdIdOp)) {
+        for (uint32_t bdId : bdIdOpToBdIdsMap[bdIdOp])
+          generator.releaseBdId(bdId);
+      } else {
+        return bdIdOp.emitOpError() << "no BD IDs found for this expression";
       }
     } else {
-      return bdIdOp.emitOpError() << "no BD IDs found for this expression";
+      // Else, must be a constant BD ID.
+      uint32_t bdId = getConstantIndexOrAssert(value);
+      generator.releaseBdId(bdId);
     }
-  } else {
-    // Else, must be a constant BD ID.
-    uint32_t bdId = getConstantIndexOrAssert(value);
-    generator.releaseBdId(bdId);
+    return success();
   }
-  return success();
+
+  /// DmaOps are assigned Bd Ids prior to invoking this function and a map
+  /// `dmaOpToBdIdMap` is maintained that maps a DmaOp to its source/target Bd
+  /// Ids. For each DmaOp in the list `dmaOps`, this API will check the
+  /// `dmaOpToBdIdMap` and release Bd Ids if assigned.
+  LogicalResult releaseAssignedBdIdsInDmaOps(
+      SmallVectorImpl<AMDAIE::NpuDmaCpyNdOp> &dmaOps) {
+    // Release BD ID used by input DMA op.
+    for (AMDAIE::NpuDmaCpyNdOp npuDmaOp : dmaOps) {
+      if (AMDAIE::BdIdOp bdIdOp = dmaOpToBdIdMap[npuDmaOp][0]; bdIdOp) {
+        if (failed(releaseBdId(bdIdOp))) return failure();
+      }
+      if (AMDAIE::BdIdOp bdIdOp = dmaOpToBdIdMap[npuDmaOp][1]; bdIdOp) {
+        if (failed(releaseBdId(bdIdOp))) return failure();
+      }
+    }
+    return success();
+  }
+
+ public:
+  BdIdAssignmentUtil(
+      DenseMap<Value, ChannelBdIdGenerator> &shimTileToGeneratorMap) {
+    this->shimTileToGeneratorMap = shimTileToGeneratorMap;
+  }
+
+  /// Assign Bd Ids to each DmaBatch belonging to a particular Tile in the
+  /// TileDmaBatch. And maintain a mapping of DmaOp -> source/target Bd Id which
+  /// would be used later during new DmaOp creation/replacement.
+  LogicalResult assignRequiredBdIdsInBatch(
+      IRRewriter &rewriter, TileDmaBatchGraph &tileDmaBatchGraph) {
+    for (auto &[tileOp, dmaBatch] : tileDmaBatchGraph.getGraph()) {
+      if (failed(processDmaBatch(rewriter, tileOp, dmaBatch))) return failure();
+    }
+    return success();
+  }
+
+  /// Traverse each DmaOp inside ControlCode and replace it with new new DmaOp
+  /// that has Bd Ids assigned using `dmaOpToBdIdMap`.
+  LogicalResult replaceDmaOps(IRRewriter &rewriter,
```

Comment:
I feel like this function belongs outside the class and you could add a helper function to the class to return the `dmaOpToBdIdMap`, ideally as a const reference, as this the only info you need inside this function.